src/Storages/StorageReplicatedMergeTree.cpp

```diff
@@ -2692,6 +2697,67 @@ void StorageReplicatedMergeTree::updateQuorum(const String & part_name)
 }
 
 
+void StorageReplicatedMergeTree::cleanLastPartNode(const String & partition_id, const String & part_name)
+{
+    auto zookeeper = getZooKeeper();
+
+    /// The name of the previous part for which the quorum was reached.
+    const String quorum_last_part_path = zookeeper_path + "/quorum/last_part";
+
+    /// Delete information from "last_part" node.
+
+    while (true)
+    {
+        Coordination::Requests ops;
+        Coordination::Responses responses;
+
+        Coordination::Stat added_parts_stat;
+        String old_added_parts = zookeeper->get(quorum_last_part_path, &added_parts_stat);
+
+        ReplicatedMergeTreeQuorumAddedParts parts_with_quorum(format_version);
+
+        if (!old_added_parts.empty())
+            parts_with_quorum.fromString(old_added_parts);
+
+        /// Delete information about particular partition.
+
+        /// Since c++20.
+        if (!parts_with_quorum.added_parts.contains(partition_id))
+        {
+            /// There is no information about interested part.
+            break;
+        }
+
+        /// De Morgan's law
+        if (part_name.empty() || parts_with_quorum.added_parts[partition_id] == part_name)
+            parts_with_quorum.added_parts.erase(partition_id);
+        else
+            break;
+
+        String new_added_parts = parts_with_quorum.toString();
+
+        auto code = zookeeper->trySet(quorum_last_part_path, new_added_parts, added_parts_stat.version);
```

Comment:
We check version because another replica can simultaneously change the same node?

Reply:
Exactly